compiler/rustc_typeck/src/collect.rs

```diff
@@ -2469,43 +2469,210 @@ fn compute_sig_of_foreign_fn_decl<'tcx>(
         None,
     );
 
-    // Feature gate SIMD types in FFI, since I am not sure that the
-    // ABIs are handled at all correctly. -huonw
-    if abi != abi::Abi::RustIntrinsic
-        && abi != abi::Abi::PlatformIntrinsic
-        && !tcx.features().simd_ffi
-    {
-        let check = |ast_ty: &hir::Ty<'_>, ty: Ty<'_>| {
-            if ty.is_simd() {
-                let snip = tcx
-                    .sess
-                    .source_map()
-                    .span_to_snippet(ast_ty.span)
-                    .map_or_else(|_| String::new(), |s| format!(" `{}`", s));
-                tcx.sess
-                    .struct_span_err(
-                        ast_ty.span,
-                        &format!(
-                            "use of SIMD type{} in FFI is highly experimental and \
-                             may result in invalid code",
-                            snip
-                        ),
-                    )
-                    .help("add `#![feature(simd_ffi)]` to the crate attributes to enable")
-                    .emit();
-            }
-        };
+    // Using SIMD types in FFI signatures requires the signature
+    // to have appropriate `#[target_feature]`'s enabled.
+    if abi != abi::Abi::RustIntrinsic && abi != abi::Abi::PlatformIntrinsic {
         for (input, ty) in iter::zip(decl.inputs, fty.inputs().skip_binder()) {
-            check(&input, ty)
+            simd_ffi_check(tcx, def_id, &input, ty)
         }
         if let hir::FnRetTy::Return(ref ty) = decl.output {
-            check(&ty, fty.output().skip_binder())
+            simd_ffi_check(tcx, def_id, &ty, fty.output().skip_binder())
         }
     }
 
     fty
 }
 
+/// Returns `Ok()` if the target-feature allows using the SIMD type on C FFI.
+/// Otherwise, returns `Err(Some())` if the target_feature needs to be enabled or
+/// or `Err(None)` if it's unsupported.
+/// Some notes about arguments:
+/// - `simd_len`: A vector register size as octet, i.e. vN in the tests.
+/// - `simd_elem_width`: A width of each element as octet, e.g. v512 should be 16 on avx512.
+fn simd_ffi_feature_check(
+    target: &str,
+    simd_width: u64,
+    simd_elem_width: u64,
+    feature: Option<String>,
+) -> Result<(), Option<&'static str>> {
+    let feature = feature.unwrap_or_default();
+    match target {
+        t if t.contains("x86") => {
+            // FIXME: this needs to be architecture dependent and
+            // should probably belong somewhere else:
+            // * on mips: 16 => msa,
+            // * wasm: 16 => simd128
+            match simd_width {
+                8 if feature.contains("mmx")
```

Comment:
Maybe instead of match:
```
if feature.contains("mmx") and simd_with == 8 {
  return Ok(());
} 
if feature.contains("see") and 8..=16.contains(simd_with) {
  return Ok(());
}
...

if feature.contains("avx512") and 8..=64.contains(simd_width) {
  return Ok(());
}
```

Reply:
Hmm, I'm not sure if it's much improvement, at least the current is readable to me. And I feel like that it makes it slightly harder to understand that what `8`, `16`, `32`, ... mean IMO.